pybind11/how-to.rst

```diff
@@ -650,105 +771,40 @@ Following :ref:`this rule <style-guide-pybind11-declare-template-wrappers>` we d
 
-Finished wrappers (C++ part)
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Cross-module dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The end results for the C++ part of the wrappers (see next for the Python part) are ``exampleTwo.cc``:
+All of the depenxencies in the example classes we've defined here are within the same compiled module, and hence we've been able to rely on ``WrapperCollection`` and its callback system to ensure the wrappers are defined in an order that works.
+The only exception is inheritance: ``ExampleTwo`` and ``ExampleThree`` both inherit from ``ExampleBase``, and that means it's critical that the ``wrapType`` call (or more precisely, the ``py::class_`` instantiation) for ``ExampleBase`` appears before that of either of its derived classes.
 
-.. code-block:: cpp
-
-    #include "pybind11/pybind11.h"
-
-    #include "lsst/TMPL/ExampleTwo.h"
+When dependencies extend beyond module boundaries, we need to import the modules that provide the classes we're using.
+While it's possible to do that directly with ``py::module::import`` calls, it's more readable and less error-prone to use the ``addInheritanceDependency`` and ``addSignatureDependency`` methods of ``WrapperCollection``.
+As its name suggests, the former is needed when you want to inherit from a base class defined in another module; this must be called before the ``py::class_`` instantiation of the derived class.
+``addSignatureDependency`` declares that the external type is used only in function or method signatures.
+Regardless of when it is called, these dependencies will be imported after all local types are declared and before any definition callbacks are run.
 
-    namespace py = pybind11;
-    using namespace pybind11::literals;
-
-    namespace lsst {
-    namespace tmpl {
+Circular *inheritance* dependencies are impossible in both C++ and Python, but circular signatures dependencies are relatively common in C++ (via forward declarations) and quite possible in Python (via function-level imports and duck-typing).
+In pybind11 wrappers, circular signature dependencies are a problem.
+Using ``WrapperCollection`` solves that problem within a single module (in which all wrappers are added to the same ``WrapperCollection`` instance), and it takes steps to solve across multiple modules, by executing signature-dependency imports between type definitions and function definitions.
+But this only solves cross-module circular dependencies to the extent that a circular Python imports result let the second importer see the types in the partially-defined first module.
+It's not clearly that something guaranteed by the CPython interpreter (let alone the Python language), so it's much safer too just make sure any types with a circular dependency are handled by the same ``WrapperCollection`` instance and defined in the same module.
```

Comment:
Remove "that".  "too" -> "to".  Is it always possible to have circularly-dependent types defined in the same module?

Reply:
Maybe not, but I think the cases where it would be impossible are sufficiently exotic that they're unlikely to ever arise in our system.  I think you'd need to have two packages with their own C++ libraries with one optional package-level dependency and one required one, and the required package would need to have Python bindings that don't depend on any of its own symbols that are provided by the optional library.